### ISSUE TITLE
Combine individual Birthday fields to a single field of format: 'date-time'

### ISSUE THREAD

**Issue by shieldz** [adobe/xdm#469]
## What are the schemas that are affected by the issue
Profile
## What are examples of products that are impacted by the issue
Today we deal with most dates by creating a field where the format is "date-time." Examples include:
Anniversary
Subscription date
Purchase date
Renewal date
Membership date

But for birthday, we have three separate fields, all of format: 'integer'
* birthDay
* birthMonth
* birthYear

Request these be deprecated in favor of a single 'date-time' field for consistency and to reduce required workarounds in downstream services (ex. special casing integer fields to support additional date-related functions.

**Comment by lrosenthol** [comment]
Huge +1 to fixing this!

**Comment by kstreeter** [comment]
The birth date fields were represented this was in order to handle cases where a complete date is not known or desired.  For example, you may only know an age (i.e. birthYear) or you may only know the day/month but not the year (birthday, without knowing how old a person is).  Any change here needs to support these other cases.

Also, procedurally this schema is "stabilizing" so we should deprecate, rather than remove, any existing fields, unless we are prepared to treat this as a critical breaking change.

Please note I am not saying that fixing/changing this is a bad idea, just that it would have to be done with care!

**Comment by shieldz** [comment]
Makes sense, Kevin.  The challenge here is that the current classification of these fields as 'integer' makes it difficult to apply date functions to these without building field-specific special cases, which we want to avoid.
Can we perhaps agree on a method for handling dates in which portions of that date (day, month, or year) are not known? That way we can support the cases you mentioned and maintain consistency as well.
We should also apply whatever decision we make here to other date fields, such as anniversary, where year may not be known or needed.

**Comment by shieldz** [comment]
Correction, the ISO standard already dictates how to handle null missing fields.  @kstreeter this means cases with incomplete dates should be manageable provided the incoming data conforms to the standard.  Adding it below for reference:


Writing Omitted Components
An omitted component can be represented by a hyphen ( - ) or an x in the extended datetime form yyyy-mm-ddThh:mm:ss and in the extended duration form Pyyyy-mm-ddThh:mm:ss.

Omitted components in the durations form PnYnMnDTnHnMnS are dropped, they do not contain a hyphen or x. For example, P2mT4H.

The following formats write omitted components that use the hyphen and the x:

Format    Datetime Form    Duration Form    Examples
$N8601H    yyyy-mm-ddThh:mm:ss    PnYnMnDTnHnMnS    --09-15T15:-:53
P2Y2DT4H5M6S/--09-15T15:-:00

$N8601EH    yyyy-mm-ddThh:mm:ss    Pyyyy-mm-ddThh:mm:ss    P000---02T02:55:20/2008---15T-:-:45
$N8601X    yyyy-mm-ddThh:mm:ss    PnYnMnDTnHnMnS    P2Y2DT4H5M6S/x-09-15T15:x:00
$N8601EX    yyyy-mm-ddThh:mm:ss    Pyyyy-mm-ddThh:mm:ss    P0003-x-02T02:55:20/2008-x-15Tx:x:45
Datetime values with omitted components that are formatted with either the $N8601B format or the $N8601BA format are formatted in the extended notation using the hyphen for omitted components to ensure accurate data. For example, when the month is an omitted component, the value 2008---15 is written and not 2008-15.

The extended notation with hyphens is also used in place of the basic notation if a duration is formatted by using the $N8601BA format. Using the same date, P2008---15 is written and not P2008-15.

**Comment by fmeschbe** [comment]
Technically a huge +1 to this request from me as well.

Unfortunately, the JSON Schema spec refers to [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6) which does not allow for missing elements because RFC 3339 defines a subset of ISO 8601 as a profile. See [JSON Schema Validation Draft 01, section 7.3.1](https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-7.3.1).

We might for sure define this for XDM to be supported, but I fear we will create interoperability problems with standard JSON Schema tooling if we go there.

How about definining birthday as a `string` field with a  regular expression match and describe it to be an RFC 3339 `full-date` with the support for missing fields ?

**Comment by trieloff** [comment]
> How about definining birthday as a string field with a regular expression match and describe it to be an RFC 3339 `full-date` with the support for missing fields ?

+1 as long as this difference is pointed out in the description.

**Comment by lrosenthol** [comment]
I have always considered date-time to be the full 8601 and not the 3339 subset...I think that's a bigger question we need to consider is which of those two we are standardizing on...

**Comment by fmeschbe** [comment]
> I think that's a bigger question we need to consider is which of those two we are standardizing on...

I think RFC 3339 is mostly sufficient. AFAICT this birthday topic is the only where RFC 3339 falls short.

**Comment by lrosenthol** [comment]
After reviewing 3339, I believe you are correct, @fmeschbe.  The only thing missing are partial dates, such as we are using here.   While I like your idea of using a string with regex, it should be as defined by 8601 instead of 3339.

**Comment by fmeschbe** [comment]
Yes, @lrosenthol, my idea is indeed to to define the regex such, that it is an extension to pure RFC 3339 (fully defined dates) but compatible with ISO 8601, so no new inventions.

**Comment by shieldz** [comment]
Okay, so the proposal is: add a standard field "birthDate" with the appropriate type. We can deprecate the old fields later.
What kind of LOE are we looking at for this addition?

**Comment by cfraser** [comment]
I've opened #481 to get a PR out there to discuss.

**Comment by kstreeter** [comment]
@fmeschbe, can you post a PR with your recommendation?  I'm not quite clear on how we would represent missing elements in a way that would be really interoperable.  As you mentioned RFC3339 doesn't support missing fields.  

The latest versions of ISO 8601 do have a provision for indicating an unspecified element, but I'm not sure how well that will work with existing tooling.  For example, if I try to load an ISO 8601 string of the form "XXXX-01-01" for a Jan 1 birthday into the Java Date class, what will happen?

I feel like it might make sense for us to make the full birth date, the birthday, and the age cases all distinct.  Then we could define birthday in the partial 8601 form without worrying about what would happen if someone tried to treat it as a real date.
